### DISCUSSION TITLE
Autocomplete conflict

### DISCUSSION THREAD

**Discussion by legendSabbir** [mkslanc/ace-linters#46] [General]
The snippets are added via plugin in Acode app . [snippet plugin](https://github.com/deadlyjack/acode-plugin-snippets/blob/main/src/main.js/) .

The issue is when i use ace-linters it blocks these custom snippets. 
also when using css the autocompletemenu 
inserts text like "..container" 2 dots . This only happens if i start typing with dot for classes .

In html file local suggestions doesn't showup
and many more small issues .

i've used webworker method . I dont know how to use it via websocket .

Anyway you can build a ace-linters plugin for the app. Acode app has a store for plugin .

[Acode plugin Docs](https://acode.foxdebug.com/plugin-docs)

If you need help for things like where to start and others i can help you . 

Many developers use mobile phone for coding and acode is a great app . It's just lacking lsp

**Comment by mkslanc** [comment]
Hey @legendSabbir , 
Lets split those issues to parts: 
1. Custom snippets were with `ace-linters` were blocked by design, but seems it's not good solution, so we will provide option to save any other completers instead of replacing them.
2. Autocompletion replacement problem (..container) seems should be related to not merged [Completions improvements](https://github.com/ajaxorg/ace/pull/5005) PR. It's planned to be merged soon, so hopefully it would be solved in next version
3. Html local suggestions is the same replaced completers issue as 1 one.

If you could provide more information about other bugs that would be really appreciated.
You could try to reproduce those issues on [demo page](https://mkslanc.github.io/ace-linters/default_services.html) it would bring more clarity for us.

I hope I answered to all questions

**Reply by legendSabbir** [reply]
(.. container) issue only occurs using ace-linters . Default ace autocomplete doesn't suggest starting with (.) and its ok like that .

check javascript linters its broken . you can check it on demo page .

it shows errors on document.querySelector() it says 'document' is not defined . Autocomplete menu doesn't even show up on javascript . For this reason i use typescript linters as javascript .
also on btn.addEventListener("") this should only suggest event listeners like click, dblclick, scroll etc and there is also the issue of ""click"" quotes note being replaced .

Syntax Highlights feature request:
  is it possible in ace to add  `Bracket pair colorization` ?


look at the autocompletemenu . It looks exactly similar to vscode and its ace's autocomplete menu with modification you can check it on `Spck editor` App on  mobile .
![Screenshot_20230331_052612](https://user-images.githubusercontent.com/123853512/228986709-05b65465-bf4d-4820-bd34-ce23714b6bee.jpg)

**Reply by mkslanc** [reply]
> check javascript linters its broken . you can check it on demo page .
> 
> it shows errors on document.querySelector() it says 'document' is not defined . Autocomplete menu doesn't even show up on javascript . For this reason i use typescript linters as javascript .

It's not broken, it just used ESlint as syntax checker and ESLint is not providing any hover information, completions and other lsp stuff - only diagnostics. To allow  browser's global object you could provide env configuration for this service, like this - 
```javascript
languageProvider.setGlobalOptions("javascript", {
    env: {
        browser: true,
        amd: true,
        builtin: true,
        node: true,
        jasmine: false,
        mocha: true,
        es6: true,
        jquery: false,
        meteor: false,
    }
});
```

Or simply use alternative typescript service, which supports all LSP features.

> also on btn.addEventListener("") this should only suggest event listeners like click, dblclick, scroll etc and there is also the issue of ""click"" quotes note being replaced .

We are relying on information that typescript service providing on cursor position. Non-replacing quotes also would be solved by `Completion Improvements PR`

> Syntax Highlights feature request:
is it possible in ace to add Bracket pair colorization ?
I would ask Ace team, if they are interested in that feature

**Reply by legendSabbir** [reply]
is there any way to only use the syntax error checks and avoid using the autocomplete feature by ace-linters . use the default autocomplete of ace ??? till things are fixed

**Reply by mkslanc** [reply]
I think I will add ability to use multiple completers and publish new version today

**Reply by legendSabbir** [reply]
ok notify me when your done . And also any workaround till `completion improvements pr` gets merged ? or how long have to wait ?

**Reply by mkslanc** [reply]
Published new version - https://www.npmjs.com/package/ace-linters/v/0.7.1
Now you could provide options for `LanguageProvider` like this - 
```javascript
let languageProvider = LanguageProvider.create(worker, {
    functionality: {
        hover: false,
        completion: false,
        completionResolve: false,
        format: false
    }
});
```  
(If you want only diagnostics)

or like this 

```javascript
let languageProvider = LanguageProvider.create(worker, {
    functionality: {
        hover: false,
        completion: {
            overwriteCompleters: false
        },
        completionResolve: false,
        format: false
    }
});
``` 
(if you want all features, but with custom completers from your side)

Couldn't think about workaround for completions except using custom ace fork. It shouldn't take long - something near 3-4 days.

**Reply by legend-sabbir** [reply]
how can i pass this functionality object if i use fromCdn ?

**Reply by mkslanc** [reply]
Similar to this [Playground example](https://mkslanc.github.io/ace-playground/?value=Ly9zYW1wbGVzL3R5cGVzY3JpcHQgbGFuZ3VhZ2Ugc2VydmljZQoKYWNlLnJlcXVpcmUoImFjZS9leHQvbGFuZ3VhZ2VfdG9vbHMiKTsKLy8gRWFzaWVzdCB3YXkgdG8gcnVuIEFDRSBlZGl0b3IgaW4gY29udGFpbmVyIHdpdGggaWQ9ImV4YW1wbGUiCmxldCBlZGl0b3IgPSBhY2UuZWRpdCgiZXhhbXBsZSIsIHsKICAgIHRoZW1lOiAiYWNlL3RoZW1lL3RleHRtYXRlIiwKICAgIG1vZGU6ICJhY2UvbW9kZS90eXBlc2NyaXB0IiwKICAgIHZhbHVlOiAiY29uc29sZS5sb2coJ0hlbGxvIHdvcmxkJyA7Igp9KTsKLy8gZW5hYmxlIGF1dG9jb21wbGV0aW9uIGFuZCBzbmlwcGV0cwplZGl0b3Iuc2V0T3B0aW9ucyh7CiAgICBlbmFibGVCYXNpY0F1dG9jb21wbGV0aW9uOiB0cnVlLAogICAgZW5hYmxlU25pcHBldHM6IHRydWUsCiAgICBlbmFibGVMaXZlQXV0b2NvbXBsZXRpb246IHRydWUKfSk7CgovL2NyZWF0ZSBMYW5ndWFnZSBQcm92aWRlciBmb3IganNvbiBkb2MgZnJvbSBhY2UgbGludGVycwpsZXQgb3B0aW9ucyA9IHsKICAgIGZ1bmN0aW9uYWxpdHk6IHsKICAgICAgICBob3ZlcjogZmFsc2UsCiAgICAgICAgY29tcGxldGlvbjogZmFsc2UsCiAgICAgICAgY29tcGxldGlvblJlc29sdmU6IGZhbHNlLAogICAgICAgIGZvcm1hdDogZmFsc2UKICAgIH0KfQpsZXQgcHJvdmlkZXIgPSBMYW5ndWFnZVByb3ZpZGVyLmZyb21DZG4oImh0dHBzOi8vY2RuLmpzZGVsaXZyLm5ldC9ucG0vYWNlLWxpbnRlcnMvYnVpbGQiLCBvcHRpb25zKTsKCi8vIFJlZ2lzdGVyIHRoZSBlZGl0b3Igd2l0aCB0aGUgbGFuZ3VhZ2UgcHJvdmlkZXIKcHJvdmlkZXIucmVnaXN0ZXJFZGl0b3IoZWRpdG9yKTsKXDBcMDwhLS0gbG9hZCBhY2UgbGFuZ3VhZ2UgdG9vbHMgLS0%2BCjxzY3JpcHQgc3JjPSJodHRwczovL2NkbmpzLmNsb3VkZmxhcmUuY29tL2FqYXgvbGlicy9hY2UvMS4xMi41L2V4dC1sYW5ndWFnZV90b29scy5taW4uanMiPjwvc2NyaXB0Pgo8IS0tIGxvYWQgYWNlIGxpbnRlcnMgLS0%2BCjxzY3JpcHQgc3JjPSJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL2FjZS1saW50ZXJzQDAuNy4xIj48L3NjcmlwdD4KPGRpdiBpZD0nZXhhbXBsZScgc3R5bGU9ImhlaWdodDogMTAwJSI%2BPC9kaXY%2BCg%3D%3D#typescript-language-service)

```javascript
let options = {
    functionality: {
        hover: false,
        completion: false,
        completionResolve: false,
        format: false
    }
}
let provider = LanguageProvider.fromCdn("https://cdn.jsdelivr.net/npm/ace-linters/build", options);
```

**Reply by legend-sabbir** [reply]
new bug

it hides the first suggestion in the autocompletemenu . it only occurs if `options` are used .

```javascript
let options = {
      functionality: {
        hover: false,
        completion: {
          overwriteCompleters: false
        },
        completionResolve: false,
        format: true
      }
    }
``` 

The first suggestion is the important one

 
you can reproduce it .
for example set the mode to `typescript`
then type p or any letter if the bug doesn't occur then backspace and agin type something . do this 2 to 3 times the bug will occur

![Screenshot_20230401_234814](https://user-images.githubusercontent.com/105509363/229306581-b1577084-fbdf-42b7-8fd7-669a46ff52f8.jpg)

**Reply by mkslanc** [reply]
Use it with `completion: false` for now (it would allow you to use any of your completers without ace-linters ones), until completers would be fixed

**Reply by mkslanc** [reply]
Seems Ace's [feat: inline autocompletion #5084](https://github.com/ajaxorg/ace/pull/5084) has some undocumented breaking changes, cause it's not reproducing in version before, I will need time to figure out what changed and correct it.

**Reply by mkslanc** [reply]
It's working for localhost with simple static server like - [http-server](https://github.com/http-party/http-server) .

**Reply by legendSabbir** [reply]
[check it](https://github.com/legendSabbir/linters/blob/master/src/main.js) .

I need help😅 . This is working fine but it needs internet access everytime . i want to make it work offline . Any idea on how can i make it . I need guide . Whatever the is inside the dist folder i can get it like , this.baseUrl + ace-linters.js .

so tried to use .fromCdn(this.baseUrl)  and i copied all file from build to dist . but it didn't worked .

this.baseUrl is equivalent of
`https://localhost/__cdvfile_files-external__/plugins/acode.plugin.test/`

**Reply by mkslanc** [reply]
1. Run static server on any port (for example 8080) for `build` dir of `ace-linters`
2. Change 7 line url to `http://localhost:8080/ace-linters.js`
3. Change 43 line url to `http://localhost:8080/`

That should be enough, I tried it on my machine without Internet with locally built `Ace Playground`

**Reply by legendSabbir** [reply]
yeah its working i checked thats why i deleted previous reply about localhost . the problem is it is not working with this url
https://localhost/__cdvfile_files-external__/plugins/acode.plugin.test/ 

but i can get files with fetch
for example
https://localhost/__cdvfile_files-external__/plugins/acode.plugin.test/ace-linters.js

`There is no problem with your library `  i just asking for tips about how can i make it work in other ways

Hey is it possible to bundle all the files ? so its one time download .
then for example i will just pass
the bundled file in from cdn or any other method ?

**Reply by mkslanc** [reply]
Bundle it as one file? It would be too big file for web ~60 mb of compressed data. 

Don't know about your restrictions, but you could use it directly, without `fromCdn` endpoint, if it's hosted on same server. Like this - 
```javascript
import {LanguageProvider} from "ace-linters/build/ace-linters";

// Create a web worker
let worker = new Worker(new URL('./webworker.js', import.meta.url));
// Create a language provider for web worker
let languageProvider = LanguageProvider.create(worker);

// Register the editor with the language provider
languageProvider.registerEditor(editor);
```

[Webworker example](https://github.com/mkslanc/ace-linters/blob/main/packages/demo/webworker-lsp/webworker.ts)

**Reply by legendSabbir** [reply]
there is issues with this .

any way if i add the script in via script tag then is there any way to connect it ?

like
```html
<script src="service-manager.js"></script>
<script src="javascript-service.js"></script>
```

then somehow register the editor 😕 .
if no way then i will use fromCdn .

**Reply by mkslanc** [reply]
What kind of issues? 

I made an example to use inline worker and fetch (I think this is what you needed to make it work with your use case) - https://github.com/mkslanc/ace-linters/blob/main/packages/demo/example-fetch-inline-worker.html

**Reply by legendSabbir** [reply]
worked like a charm. thanks man . You helped a lot.

**Reply by legendSabbir** [reply]
is inline-autocomplete available ? or in development ?

if available how can i test it .


also is there any way to get file path suggestion in ace-linters ?😕

**Reply by mkslanc** [reply]
New version with inline-autocomplete haven't been published yet, I think it would be done in near future. With file path suggestion - it's tricky,  for now there is no such functionality, cause we are relying on web and in web all file API is experimental at the moment. (and most of it locked due to security reasons).

**Reply by mkslanc** [reply]
You could try to pass filenames and content to `globalOptions` like this - 
```javascript
        languageProvider.setGlobalOptions("typescript", {
            extraLibs: {
                "path/filename": {
                    content: contentOfYourFile,
                    version: 1
                },
            },
            compilerOptions: {
                allowJs: true,
                checkJs: true
            }
        }, true);
```

**Reply by legendSabbir** [reply]
hey . someone reported me that php formatter is not working .

```php
<?php

class Test 
{
             public function test(     )
  {
    
  }
  
}
class TestTwo 
{
  public function __construct(public AnotherController $controller, public AnotherController2 $controller2, public AnotherController3 $controller3)
             {
 
    
  }
  public function test2()
  {
    $getSomething = $controller3->something;
                             $fetch = $controller->something;
    return $fetch.$getSomething;
    
  }
}
```

he gave me this code . The formatter is not working . Also javascript formatter removes spaces between "="

```javascript
const abc = "Hello world"
// Formatted into
const abc="Hello world"
// this is unwanted 
```

**Reply by mkslanc** [reply]
Yeah, there is no free PHP web service with formatter/completers/hover. So we used corrected `php.js 0.1.0` parser the same one used in ace and it has only diagnostics.  For full functional LSP support you could use `AceLanguageClient` and server built using websockets. You could look at example - https://github.com/mkslanc/ace-linters/tree/main/packages/demo/websockets-lsp

**Reply by mkslanc** [reply]
For formatter settings - it's default ts ones with support of ace's options like `options.insertSpaces,` and `options.tabSize` . We could provide ability for user to set their own one, similar to `completerOptions` for ts. Could you write feature request in Issues for that?

**Reply by legendSabbir** [reply]
people also like this ?
```javascript
const abc="Hello world"
length-1
// etc 
```
also is it fixable ?

**Reply by mkslanc** [reply]
Yeah, that property should be `true` for that behaviour - https://github.com/mkslanc/ace-linters/blob/main/packages/ace-linters/services/typescript/typescript-service.ts#L159

I am not sure, if this should be default state, anyway we need a way to allow users to pass custom formatter options.

**Reply by legendSabbir** [reply]
i've tested with `true` for that property and it is working . Also `css` has this same issue

```css
ul > li {}

/* Turns into */
ul>li {}
```

any property to change for css ?
(edit)
I found it in the compiled code .

also why formatter not working for python ?

**Reply by mkslanc** [reply]
Python also has only error checking. The idea of Ace linters mostly in providing tools to build your own services and connect with editor, not to cover all possible languages :)

**Reply by mkslanc** [reply]
With web-workers we are restricted by not so many services implementation. Most of them are native, so we could use them only via WebSockets or if someone compiled libs to wasm

**Reply by legendSabbir** [reply]
![Screenshot_20230410_181349](https://user-images.githubusercontent.com/123853512/230899927-2e850921-7b44-431a-b0cb-17804294fc16.jpg)
![Screenshot_20230410_181516](https://user-images.githubusercontent.com/123853512/230899955-16bcc831-39ca-42f1-9b26-47a743ee14ea.jpg)

There is many more errors like this what can i do about them ?

while the mode is typescript

**Reply by mkslanc** [reply]
They all are solving with service config - 
for first one you need content of additional typescript module "fs" to add to `extraLibs` property and set `moduleResolution` to 2, 3 or 99 (`ModuleResolutionKind.NodeJs`, `ModuleResolutionKind.Node12`, `ModuleResolutionKind.NodeNext`)  similar to `tsconfig.json` one's
```javascript
languageProvider.setGlobalOptions("typescript", {
    compilerOptions: {
        allowJs: true,
        target: ScriptTarget.ESNext,
        jsx: JsxEmit.Preserve,
        moduleResolution: 3
    },
    extraLibs: {
        "fs.lib.ts": {
            content: fsModule, //content of node_modules/@types/node/fs.d.ts
            version: 0
        }
    }
});
```

for second one you need to set `module` property to ES version >=es2020 -
```javascript
languageProvider.setGlobalOptions("typescript", {
    compilerOptions: {
        module: 6 //This is ModuleKind.ES2020
    }
}, true);
```

![image](https://user-images.githubusercontent.com/32402726/231151347-e0a4f59b-8d3a-46c4-93ec-9ee094c322c9.png)


So, all those kind of problems are solving by configuration, which could be different for each project - similar to [`tsconfig.json`](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html)

**Reply by legendSabbir** [reply]
There were more errors .

So basically how can i disable the error checks for `tsconfig.json`  and other `@types` check ? 
because its problematic .

**Reply by mkslanc** [reply]
It's a pity, but there is still no option to exclude different rules in Typescript itself, I could provide some way to filter rules, but I don't know all rules I need to exclude in this case. So you will need to form your own filter set of rules.

**Reply by legendSabbir** [reply]
do you have any idea on how can i create a ctrl key button for mobile users so that user can execute editor commands easily ?
I tried many solutions but failed .

for example 
![Screenshot_20230418_162354](https://user-images.githubusercontent.com/123853512/232749149-bb081e57-12fb-4698-88f0-9720f952ec20.jpg)

when this key is active i want to then execute commands for ctrl-c ctrl-v etc etc for
copylinesdown
removeline
movelinesup
you get it

**Reply by mkslanc** [reply]
Need to check how it work in mobile to understand what the problem is. Meanwhile new release come with many fixes (https://github.com/mkslanc/ace-linters/releases/tag/v0.9.0)

+ Ace published new version with autocompletion fixes
+ You could manually set which features you want to use from each service

**Reply by legendSabbir** [reply]
- so the (..container) fixed now ? .
- give me some examples for manual implementations for each services
- also check the ctrl key with virtual button for mobile on free time
-  also check the typescript service . for example use the `server.ts` file and see how many error message it shows . Since its browser based thats why i want avoid `tsconfig.json` and other built in module Cheks

**Reply by mkslanc** [reply]
- First suggestion not showing up fixed (not sure if .container is the same one)
- just add `features: {completion: false, completionResolve: false, diagnostics: true, format: false, hover: false, documentHighlight: false, signatureHelp: false},` to object inside `registerService` - https://github.com/mkslanc/ace-linters/blob/v0.9.1/packages/demo/example-fetch-inline-worker.html#L88
Another example is any service from [here](https://github.com/mkslanc/ace-linters/blob/v0.9.1/packages/demo/webworker-lsp/webworker.ts)

**Reply by legendSabbir** [reply]
`Parsing error: 'import' and 'export' may appear only with 'sourceType: module'`

javascript service error annotation

**Reply by mkslanc** [reply]
> `Parsing error: 'import' and 'export' may appear only with 'sourceType: module'`
> 
> javascript service error annotation

Could you provide more details?

**Reply by legendSabbir** [reply]
```javascript
import plugin from '../plugin.json';
const { editor } = editorManager 

class AcodePlugin {
  async init() {
    const quickTools = document.querySelector("#quick-tools")
    const mut_observer = new MutationObserver((mutations) => {
      mutations.forEach((mutation) => {
        const shiftKey = mutation.target.querySelector("#shift-key")
        
        mut_observer.disconnect();
      });
    });
    
    mut_observer.observe(quickTools, {childList: true});
  }
  
  async destroy() {
  }
}

if (window.acode) {
  const acodePlugin = new AcodePlugin();
  acode.setPluginInit(plugin.id, (baseUrl, $page, { cacheFileUrl, cacheFile }) => {
    if (!baseUrl.endsWith('/')) {
      baseUrl += '/';
    }
    acodePlugin.baseUrl = baseUrl;
    acodePlugin.init($page, cacheFile, cacheFileUrl);
  });
  acode.setPluginUnmount(plugin.id, () => {
    acodePlugin.destroy();
  });
}

// Errors 
[
    {
        "row": 0,
        "column": 19,
        "text": "Cannot find module '../plugin.json'. Consider using '--resolveJsonModule' to import module with '.json' extension.",
        "type": "error"
    },
    {
        "row": 1,
        "column": 19,
        "text": "Cannot find name 'editorManager'.",
        "type": "error"
    },
    {
        "row": 8,
        "column": 41,
        "text": "Property 'querySelector' does not exist on type 'Node'.",
        "type": "error"
    },
    {
        "row": 21,
        "column": 11,
        "text": "Property 'acode' does not exist on type 'Window & typeof globalThis'.",
        "type": "error"
    },
    {
        "row": 23,
        "column": 2,
        "text": "Cannot find name 'acode'.",
        "type": "error"
    },
    {
        "row": 27,
        "column": 16,
        "text": "Property 'baseUrl' does not exist on type 'AcodePlugin'.",
        "type": "error"
    },
    {
        "row": 28,
        "column": 21,
        "text": "Expected 0 arguments, but got 3.",
        "type": "error"
    },
    {
        "row": 30,
        "column": 2,
        "text": "Cannot find name 'acode'.",
        "type": "error"
    },
    {
        "row": 0,
        "column": 0,
        "text": "Parsing error: 'import' and 'export' may appear only with 'sourceType: module'",
        "type": "error"
    }
]
```

**Reply by mkslanc** [reply]
With [new version](https://www.npmjs.com/package/ace-linters/v/0.10.0?activeTab=versions) you could set which errors you want to ignore or replace as warnings/info - 
For Typescript Service it would be -
```typescript
languageProvider.setGlobalOptions("typescript", {
    errorCodesToTreatAsWarning: [
        "2540"
    ],
    errorCodesToIgnore: [
        "error code", "second-error-code"
    ],
    errorCodesToTreatAsInfo: [
        "error-code"
    ]
});
```

I am closing this discussion as it is already too big one, if you would have any other request/questions, please open new one

**Comment by nightwing** [comment]
@legend-sabbir there is no public api in ace, you need to use a hack similar to
```js
editor.onTextInput =  function(text) {
    var keys = require('ace/lib/keys')
    editor.onCommandKey({}, keys.KEY_MODS["cmd"] , keys[text] || text.charCodeAt(0))
    delete editor.onTextInput
}
```

For general questions about ace, like this, or for feature request like support for Bracket pair colorization, https://github.com/ajaxorg/ace/discussions is a better place.
